// @neorecrut/shared-avatar/src/constants/index.ts
export const AVATARS = [
  {
    avatar_id: "Pedro_Blue_Shirt_public",
    name: "Pedro",
    image: "/avatars/Pedro.png",
  },
  {
    avatar_id: "Marianne_Red_Suit_public",
    name: "Marianne",
    image: "/avatars/Marianne.png",
  },
  {
    avatar_id: "Katya_Pink_Suit_public",
    name: "Katya",
    image: "/avatars/Katya.png",
  },
  {
    avatar_id: "Graham_Black_Shirt_public",
    name: "Graham",
    image: "/avatars/Graham.png",
  },
  {
    avatar_id: "Anthony_Black_Suit_public",
    name: "Anthony",
    image: "/avatars/Anthony.png",
  },
  {
    avatar_id: "Alessandra_Black_Suit_public",
    name: "Alessandra",
    image: "/avatars/Alessandra.png",
  },
  {
    avatar_id: "Anastasia_Grey_Shirt_public",
    name: "Anastasia",
    image: "/avatars/Anastasia.png",
  },
  {
    avatar_id: "Amina_Black_Suit_public",
    name: "Amina",
    image: "/avatars/Amina.png",
  },
  {
    avatar_id: "Thaddeus_Black_Shirt_public",
    name: "Thaddeus",
    image: "/avatars/Thaddeus.png",
  },
];

export const STT_LANGUAGE_LIST = [
  { label: "English", value: "en", key: "en" },
  { label: "Français", value: "fr", key: "fr" },
  { label: "Deutsch", value: "de", key: "de" },
  { label: "Italiano", value: "it", key: "it" },
  { label: "Português", value: "pt", key: "pt" },
  { label: "Español", value: "es", key: "es" },
];

export const DEFAULT_USER_ID = "00000000-0000-0000-0000-000000000001";
